File: CodingTest/LeetCode/35-search-insert-position.js
/**
 * https://leetcode.com/submissions/detail/603045481/
 * 절반씩 짤라가며 나아간다. 인덱스는 계속해서 진행 될 때 추가한다.
 * (오른쪽으로 갈 때 추가)
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function (nums, target) {
    if (nums[0] >= target) {
        return 0;
    }
    if (nums[nums.length - 1] < target) {
        return nums.length;
    }
    if (nums[nums.length - 1] === target) {
        return nums.length - 1;
    }
    let result = 1;
    while (nums.length > 1) {
        const pivot = Math.round(nums.length / 2);
        if (nums[pivot] < target) {
            nums = nums.slice(pivot, nums.length);
            result += pivot;
        } else {
            nums = nums.slice(0, pivot)
        }
    }
    return result
};

File: CodingTest/LeetCode/35-search-insert-position.java
/*
 * Binary Search 를 이용한 해결 방법으로 구현
 * 이전 js 버전보다 100배 빨라졌다. (이전 100 ms -> 이후 1m)
  */
class Solution {
    public int searchInsert(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        int mid = 0;

        if(nums[left] >= target){
            return 0;
        }

        if(nums[right] < target){
            return right+1;
        }

        while(left<=right){
            mid = left + (right - left) / 2;
            System.out.println(mid);
            if(nums[mid] == target){
                return mid;
            }
            if(nums[mid] > target){
                right = mid - 1;
            }else{
                left = mid + 1;
            }

        }

        return left;
    }
}